### DISCUSSION TITLE
Real-world hotscript use cases?

### DISCUSSION THREAD

**Discussion by gvergnaud** [gvergnaud/HOTScript#65] [Q&A]
Hey everyone!

Here is a thread to collect real-world use cases for hotscript :)

As we stabilize the API, we are very curious about the problems you are solving with this library, and if/how it simplifies your type-level logic.

Even if you are not using hotscript yet, what is the most complex type you have ever written?

**Comment by markerikson** [comment]
"Most complex types":

- The `MergeParameters` type in Reselect 4.1.5: https://github.com/reduxjs/reselect/blob/v4.1.5/src/types.ts#L100-L148  (which later broke in TS 4.9, and Anders Hjelsberg himself provided a simpler implementation that only works with TS4.7+: https://github.com/reduxjs/reselect/blob/v4.1.7/src/versionedTypes/ts47-mergeParameters.ts )
- The `connect` types in React-Redux v8 (which admittedly were ported from the v7 `@types/react-redux` package): https://github.com/reduxjs/react-redux/blob/v8.0.2/src/components/connect.tsx#L251-L407
- And then while _I_ didn't write them, basically everything in Redux Toolkit :)  [`createAction.ts`](https://github.com/reduxjs/redux-toolkit/blob/v1.9.3/packages/toolkit/src/createAction.ts), [`createSlice.ts`](https://github.com/reduxjs/redux-toolkit/blob/v1.9.3/packages/toolkit/src/createSlice.ts), [`apiTypes.ts`](https://github.com/reduxjs/redux-toolkit/blob/v1.9.3/packages/toolkit/src/query/apiTypes.ts), [`endpointDefinitions.ts`](https://github.com/reduxjs/redux-toolkit/blob/v1.9.3/packages/toolkit/src/query/endpointDefinitions.ts), etc

**Comment by gvergnaud** [comment]
Thanks for sharing @markerikson! Do you have some type-level unit tests for this `MergeParameters` function? I'd like to try reimplementing it with hotscript, but it would be much easier with a few test cases to make sure edge cases are covered

**Reply by markerikson** [reply]
There's no "unit" tests for `MergeParameters`, but we do have a `typetests` file that tests the behavior of `createSelector` as a whole:

- https://github.com/reduxjs/reselect/blob/v4.1.7/typescript_test/test.ts

Could probably reverse-engineer some "unit test" cases from that, the comments in the original complicated implementation in 4.1.5, and the discussion in https://github.com/microsoft/TypeScript/pull/50831#issuecomment-1253830522 .

**Reply by gvergnaud** [reply]
Thanks! 

I didn't test this against your suite of unit tests, but I gave this a quick try and here is what I came up with:

```ts
import * as H from "hotscript";

type res = MergeParameters<
  // ^? [21, 2, symbol, { a: number } & { b: string }]
  [
    (a: number, b: number) => boolean,
    (a: 21, b: number) => string,
    (a: number | string, b: 2, c: symbol, d: { a: number }) => string,
    (a: number | string, b: 2, c: symbol, d: { b: string }) => string
  ]
>;

type MergeParameters<T extends readonly UnknownFunction[]> = H.Pipe<
  T,
  [
    H.Tuples.Map<H.Functions.Parameters>,
    PadWithUnknown,
    Transpose,
    H.Tuples.Map<H.Tuples.ToIntersection>
  ]
>;

/** 
 * `Call<Transpose, [[1, 2], ['a', 'b']]>` // [[1, 'a'], [2, 'b']]
 */
interface Transpose extends H.Fn {
  return: H.Apply<H.Tuples.Zip, this['arg0']>
}


/**
 * Make sure all lists of arguments have the same length by 
 * adding unknown arguments at the end of the shorter ones.
 * @example
 * ```ts
 * type T = Call<PadWithUnknown, [[], [1], [1, 2]]>>
 * // [[unknown, unknown], [1, unknown], [1, 2]]
 * ```
 */
interface PadWithUnknown extends H.Fn {
  return: this["args"] extends [infer argsList extends any[][]]
    ? H.Pipe<argsList, [
        GetMaxLength,
        GetUnknownPadding<argsList>,
        H.Tuples.ZipWith<H.Tuples.Concat, argsList, H._>
      ]>
    : never
}


type GetMaxLength = H.ComposeLeft<[
  H.Tuples.Map<H.Tuples.Length>,
  H.Tuples.Reduce<H.Numbers.Max, 0>
]>

/**
 * Returns a list of `unknown` to pad the argsList
 * to make them all have the same length.
 * @example
 * ```ts
 * type T = Call<GetUnknownPadding<[[], [1], [1, 2]]>, 2>
 * // [[unknown, unknown], [unknown], []]
 * ```
 */
interface GetUnknownPadding<argsList extends any[][]> extends H.Fn {
  return: H.Eval<
    H.Tuples.Map<
      H.ComposeLeft<
        [
          H.Tuples.Length,
          H.Numbers.Sub<this['arg0'], H._>,
          H.Tuples.Range<0>,
          H.Tuples.Map<H.Constant<unknown>>,
          H.Tuples.Tail
        ]
      >,
      argsList
    >
  >
}

/** Any function with arguments */
type UnknownFunction = (...args: any[]) => any;
```


[Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wBJwGZQiOAIgAsIZUBjKYMGQgbgChGYBPMAUzig-QF44AWQ5QA5hwAKyKMhAcYI1AB5GcOAHp1cAHoB+VXADaBtQApkALjgA7AK4gARiIA0cB1buORASjh8AfG4QEAA2HMjWziZwphRWAEwAjK7uNvZOUL4BcKgw1NaiUWpmcWleUHAAPjl5wAUpCa4AJlaorI6hrhxWAN4oHukicAC+WYG5+YXRsQPlVTWTDXDxza3tDp1w3XB9qRN1oiNjCwcGALqM-kws7FzCYpLSsvKKSgAqWwAeCtZN6DzIJoQawhVhwACq1gA1tYIAB3awAMVs1goMGAwMMZ0CAgwADoJDQOCo1G8ikZovi3rYwGFUHjBMgwEp8cjUejgfSpDI5AooKh-OS1FImgB1YAwYiQmHwyLRN4yayoSCoDhCzB46m03gMpkszU0umaiAASWsfNVaIx1n850u10YdT52GQFC4Coiyogqq+Pz+GsR1h2Bh4MFsUGsVnxAEEwLTWPqtUaAFo0VyS4CoQwAcmkogADNnsYxhswnSIXW64CLxZLpbCEb6OL90Kygz0Q-Jw5G4Bms4Q86hCGcmy2jHVsENBwAZTPwDjfZv+iKsLFYi7FOC6DWEzhKGdz1zGTebgDi8kZn2nzdEkvVZ-k9dlIqaB33YlQs9yguim6php1VMwFrYhEwA+kAGFgQoZAYFcA9clcfEAH1bRPOBi03DwOAANxEEtmBuTg4HPGBL2vApJT8DUoPAb0OGvbAYCUY8NSTHVGWZf9tXpCjb2IH81G4o0ACUOCaWw3X1AA5QZ+V1T5XHzW1i0YdQEAQVQkDEsMI3QZA4BCOc4AgbA4AAAxRGUEXM3sIDgMBAV7YguAQmAtLsuAQGQKEuElDgCGQEIQjgYhkDw5yuFQZ5DJvSU8Q8gABBdZG1DzzIy8gPLYYiPgECCgpCJRSKfBEXzfQwsSPRIzmq1x4jObF6rQpBNCMQwrIbSI4E62VaqMXqEX69d0oyrT1Edc0K1dLgSuhLryoKd9RE-YyFz9fTrFXM4sUCdal1bPFA2DNRQ27KM8QAURwoKSWKYSOL1X8aPwFUGI4Ji7vQ1j0KEg0eLxPi72ev88Vk8p6QAZVsBwlD7HM80LfqUME36-vY+kRIicQlGU+8Twe+lOP1KClRgCJmMGm1UbRwnNWQYAQhBjDnpp4o3OiNC4FtUs1I0uBoy2nAUStYE4DhCViBQMR7GbchEAmnKuFKpERY5IMBFMPFtcHKwVyxY4VwdIA)

Curious to know what you think!

**Reply by markerikson** [reply]
_Looks_ plausible at first glance :)  I'll have to look at it later.

Out of curiosity, what sort of TS version range support does `hotscript` have overall?  How far back does any of this stuff work?

**Reply by gvergnaud** [reply]
We haven't thought too much about retro compatibility yet. It looks like the current version of the code is only compatible with [TSv4.8 or above](https://www.typescriptlang.org/play?ts=4.8.4#code/JYWwDg9gTgLgBAKjgQwM5wBJwGZQiOAIgAsIZUBjKYMGQgbgChGYBPMAUzig-QF44AWQ5QA5hwAKyKMhAcYI1AB5GcOAHp1cAHoB+VXADaBtQApkALjgA7AK4gARiIA0cB1buORASjh8AfG4QEAA2HMjWziZwphRWAEwAjK7uNvZOUL4BcKgw1NaiUWpmcWleUHAAPjl5wAUpCa4AJlaorI6hrhxWAN4oHukicAC+WYG5+YXRsQPlVTWTDXDxza3tDp1w3XB9qRN1oiNjCwcGALqM-kws7FzCYpLSsvKKSgAqWwAeCtZN6DzIJoQawhVhwACq1gA1tYIAB3awAMVs1goMGAwMMZ0CAgwADoJDQOCo1G8ikZovi3rYwGFUHjBMgwEp8cjUejgfSpDI5AooKh-OS1FImgB1YAwYiQmHwyLRN4yayoSCoDhCzB46m03gMpkszU0umaiAASWsfNVaIx1n850u10YdT52GQFC4Coiyogqq+Pz+GsR1h2Bh4MFsUGsVnxAEEwLTWPqtUaAFo0VyS4CoQwAcmkogADNnsYxhswnSIXW64CLxZLpbCEb6OL90Kygz0Q-Jw5G4Bms4Q86hCGcmy2jHVsENBwAZTPwDjfZv+iKsLFYi7FOC6DWEzhKGdz1zGTebgDi8kZn2nzdEkvVZ-k9dlIqaB33YlQs9yguim6php1VMwFrYhEwA+kAGFgQoZAYFcA9clcfEAH1bRPOBi03DwOAANxEEtmBuTg4HPGBL2vApJT8DUoPAb0OGvbAYCUY8NSTHVGWZf9tXpCjb2IH81G4o0ACUOCaWw3X1AA5QZ+V1T5XHzW1i0YdQEAQVQkDEsMI3QZA4BCOc4AgbA4AAAxRGUEXM3sIDgMBAV7YguAQmAtLsuAQGQKEuElDgCGQEIQjgYhkDw5yuFQZ5DJvSU8Q8gABBdZG1DzzIy8gPLYYiPgECCgpCJRSKfBEXzfQwsSPRIzmq1x4jObF6rQpBNCMQwrIbSI4E62VaqMXqEX69d0oyrT1Edc0K1dLgSuhLryoKd9RE-YyFz9fTrFXM4sUCdal1bPFA2DNRQ27KM8QAURwoKSWKYSOL1X8aPwFUGI4Ji7vQ1j0KEg0eLxPi72ev88Vk8p6QAZVsBwlEu74ZDRJQ+xzPNC36zwMkFDVUPvE8HvpESInEJRlLx0H2PpTj9SgpUYAiZjBptQTfr+ynNWQYAQhBjDnpZzc3OiNC4FtUs1I0uBoy2nAUStYE4DhCViBQMR7GbchEAmnKuFKpFZY5IMBFMPETcHKwVyxY4VwdIA) but I don't see why we couldn't make it compatible with 4.5+ (tail call optimization is a hard requirement)

**Reply by markerikson** [reply]
Gotcha. Yeah, as a point of comparison I'm currently _trying_ to support TS 4.2+ in Reselect.

**Reply by markerikson** [reply]
@gvergnaud Had a follow-up question on this.

Right now one of the limitations in Reselect is that we apparently can't carry a generic type through our internal mapped types:

- https://github.com/reduxjs/reselect/issues/571

According to @phryneas , this is a TS limitation - https://stackoverflow.com/questions/74841157/preserving-generics-in-typescript-mapped-types is probably the same kind of thing.

Any chance Hotscript makes something like that feasible?

**Comment by audunolsen** [comment]
On the more trivial side, but…

APIs w/ either inconsistent or undesirable casing in their responses.

I usually put hotscript's pipe w/ a recursive camelise function in my zod schema's last transform call to enforce camelcase w/ correct typings.

**Comment by boyum** [comment]
The Infer* types in h5p-types 

- https://github.com/boyum/h5p-types/blob/main/src/types/InferParamsFromSemantics.ts
- https://github.com/boyum/h5p-types/blob/main/src/types/InferParamTypeFromFieldType.ts
- https://github.com/boyum/h5p-types/blob/main/src/types/InferGroupParams.ts

I haven't had a real look, but I'm sure that these could be simplified by HotScript, especially since deep semantics easily breaks TS' max recursion. These tests are not too far away from the limit: https://github.com/boyum/h5p-types/blob/main/test/InferParamsFromSemantics.test.ts
